### PR TITLE
fix and feat

### LINKED ISSUE #31: Segmentation fault in http mode
Running on k2p padavan
xfrpc: 2.1.606

config:
```
[common]
server_addr = 192.168.2.2
server_port = 7000
token = ***

[Router-747d245360cd]
type = http
local_port = 80
custom_domains = 747d245360cd.frp.example.com
```

logs:
```console
K2P:/tmp # ./xfrpc -c /etc/storage/frpc.ini -f -d 7
[7][Mon Jan 23 16:01:48 2023][6940](config.c:328) Reading configuration file '/etc/storage/frpc.ini'
[7][Mon Jan 23 16:01:48 2023][6940](config.c:95) Section[common]: {server_addr:192.168.2.2, server_port:7000, auth_token:***, interval:30, timeout:90}
[7][Mon Jan 23 16:01:48 2023][6940](config.c:120) Proxy service 0: {name:Router-747d245360cd, local_port:80, type:http}
[7][Mon Jan 23 16:01:48 2023][6940](login.c:104) working in router
[6][Mon Jan 23 16:01:48 2023][6940](control.c:653) connect server [192.168.2.2:7000]...
[7][Mon Jan 23 16:01:48 2023][6940](control.c:615) xfrp server connected
[7][Mon Jan 23 16:01:48 2023][6940](control.c:690) send plain msg ----> [o: { "version": "0.43.0", "hostname": "", "os": "Linux", "arch": "mips", "user": "", "privilege_key": "0f35985cc07ead44460f1a54f37ce1de", "timestamp": 1674460908, "run_id": "747D245360CD", "pool_count": 1, "metas": null }]
[7][Mon Jan 23 16:01:48 2023][6940](control.c:627) start keep_control_alive
[7][Mon Jan 23 16:01:48 2023][6940](login.c:129) xfrp login response: run_id: [747D245360CD], version: [0.45.0]
[3][Mon Jan 23 16:01:48 2023][6940](control.c:445) login success! login_len 67 len 76 ilen 0
[7][Mon Jan 23 16:01:48 2023][6940](control.c:317) recv eas1238 iv data
[6][Mon Jan 23 16:01:48 2023][6940](control.c:159) Start xfrp proxy services ...
[7][Mon Jan 23 16:01:48 2023][6940](control.c:790) control proxy client: [Type 112 : proxy_name Router-747d245360cd : msg_len 279]
[7][Mon Jan 23 16:01:48 2023][6940](control.c:128) new client through tcp mux: 5
[7][Mon Jan 23 16:01:48 2023][6940](control.c:690) send plain msg ----> [w: { "run_id": "747D245360CD" }]

.........

[7][Mon Jan 23 16:08:46 2023][6997](client.c:182) free client 103
[7][Mon Jan 23 16:08:46 2023][6997](control.c:398) proxy service [Router-747d245360cd] [(null):80] start work connection. remain data length 0
[7][Mon Jan 23 16:08:46 2023][6997](client.c:137) proxy server [192.168.2.2:-1] <---> client [127.0.0.1:80]
[7][Mon Jan 23 16:08:46 2023][6997](client.c:78) what [128] client [105] connected : Operation now in progress
[7][Mon Jan 23 16:08:46 2023][6997](control.c:128) new client through tcp mux: 113
[7][Mon Jan 23 16:08:46 2023][6997](control.c:690) send plain msg ----> [w: { "run_id": "747D245360CD" }]
[7][Mon Jan 23 16:08:46 2023][6997](control.c:128) new client through tcp mux: 115
[7][Mon Jan 23 16:08:46 2023][6997](control.c:690) send plain msg ----> [w: { "run_id": "747D245360CD" }]
[7][Mon Jan 23 16:08:46 2023][6997](control.c:398) proxy service [Router-747d245360cd] [(null):80] start work connection. remain data length 0
[7][Mon Jan 23 16:08:46 2023][6997](client.c:137) proxy server [192.168.2.2:-1] <---> client [127.0.0.1:80]
[7][Mon Jan 23 16:08:46 2023][6997](client.c:78) what [128] client [107] connected : Operation now in progress
[7][Mon Jan 23 16:08:46 2023][6997](client.c:72) xfrpc proxy close connect server [(null):80] stream_id 107: Operation now in progress
[7][Mon Jan 23 16:08:46 2023][6997](tcpmux.c:266) free stream 107
[7][Mon Jan 23 16:08:46 2023][6997](client.c:182) free client 107
[7][Mon Jan 23 16:08:46 2023][6997](client.c:72) xfrpc proxy close connect server [(null):80] stream_id 105: Operation now in progress
[7][Mon Jan 23 16:08:46 2023][6997](tcpmux.c:266) free stream 105
[7][Mon Jan 23 16:08:46 2023][6997](client.c:182) free client 105
Segmentation fault
```

### PR DESCRIPTION
**Pull request by JimLee1996**
1. fix #31
2. adapt ifname for padavan

### CODE REVIEW

**Comment by helintongh** [comment]
only excute:
git rebase HEAD~1 --signoff and git push --force-with-lease origin fix_tcpmux_segfault

Your PR only then can be merged

**Comment by JimLee1996** [comment]
2 PRs focus on different senorios, this one is for feat and another one is for bug fix. So, is it ok to put them together?

**Comment by JimLee1996** [comment]
@helintongh Done. Please check. I'll close another pr.

**Comment by helintongh** [comment]
> @helintongh Done. Please check. I'll close another pr.

No rebase operation performed,bro.
图片
View the details of the DOC, and modify it according to the details before it can be merged

**Comment by JimLee1996** [comment]
done. please check again

**Comment by helintongh** [comment]
> done. please check again

Reported to maintainer.will be merged soon.

File: tcpmux.c [hunk 1]
```diff
@@ -331,8 +331,16 @@ static int
...
 				void (*fn)(uint8_t *, int, void *), void *param)
 {
+    if(!stream){
+        return 0;
+    }
+	uint32_t id = stream->id;
+
 	if (!process_flags(flags, stream)) return 0;
 
+	if(!get_stream_by_id(id)){
```
Comment (liudf0716) [inline tcpmux.c:341]: why need to check it again? in what condition if stream is not NULL but still cannot find it by its id?

Comment (JimLee1996) [reply tcpmux.c:341]: `process_flags()` may call `del_proxy_client_by_stream_id()`
https://github.com/liudf0716/xfrpc/blob/39b5d81231b01fae55359fb25d80ba91e385df72/tcpmux.c#L267

and `del_proxy_client` will free underlying stream struct, then the stream points to an invalid address.
https://github.com/liudf0716/xfrpc/blob/39b5d81231b01fae55359fb25d80ba91e385df72/client.c#L184

**Review by liudf0716 — APPROVED** [review]

**Comment by liudf0716** [comment]
@JimLee1996 merged, thanks!

**Pull request merged**